src/hardware/gus.cpp

```diff
@@ -1,924 +1,1418 @@
 	}
-	Bit8u ReadPanPot(void) {
-		return PanPot;
+	// Otherwise, restart the position back to its start or end
+	else {
+		ctrl.state |= 1; // Stop the voice
+		ctrl.pos = (ctrl.state & CTRL::DECREASING) ? ctrl.start : ctrl.end;
 	}
-	void WriteRampCtrl(Bit8u val) {
-		Bit32u old=myGUS.RampIRQ;
-		RampCtrl = val & 0x7f;
-		//Manually set the irq
-		if ((val & 0xa0)==0xa0) 
-			myGUS.RampIRQ|=irqmask;
-		else 
-			myGUS.RampIRQ&=~irqmask;
-		if (old != myGUS.RampIRQ) 
-			CheckVoiceIrq();
+	return;
+}
+
+bool Voice::Is8Bit() const
+{
+	return !(wave_ctrl.state & CTRL::BIT16);
+}
+
+float Voice::GetSample(const uint8_t *ram)
+{
+	const int32_t pos = PopWavePos();
+	const auto addr = pos / WAVE_WIDTH;
+	const auto fraction = pos & (WAVE_WIDTH - 1);
+	const bool should_interpolate = wave_ctrl.inc < WAVE_WIDTH && fraction;
+	float sample = Is8Bit() ? Read8BitSample(ram, addr)
+	                        : Read16BitSample(ram, addr);
+	if (should_interpolate) {
+		const auto next_addr = addr + 1;
+		const float next_sample = Is8Bit() ? Read8BitSample(ram, next_addr)
+		                                   : Read16BitSample(ram, next_addr);
+		sample += (next_sample - sample) *
+		          static_cast<float>(fraction) * WAVE_WIDTH_INV;
 	}
-	INLINE Bit8u ReadRampCtrl(void) {
-		Bit8u ret=RampCtrl;
-		if (myGUS.RampIRQ & irqmask) ret |= 0x80;
-		return ret;
+	assert(sample >= AUDIO_SAMPLE_MIN && sample <= AUDIO_SAMPLE_MAX);
+	return sample;
+}
+
+float Voice::GetVolScalar(const float *vol_scalars)
+{
+	// Unscale the volume index and check its bounds
+	const auto i = static_cast<size_t>(
+	        ceil_sdivide(PopVolPos(), VOLUME_INC_SCALAR));
+	assert(i < VOLUME_LEVELS);
+	const float scalar = vol_scalars[i];
+	return scalar;
+}
+
+void Voice::GenerateSamples(float *stream,
+                            const uint8_t *ram,
+                            const float *vol_scalars,
+                            const AudioFrame *pan_scalars,
+                            const int requested_frames)
+{
+	if (vol_ctrl.state & wave_ctrl.state & CTRL::DISABLED)
+		return;
+
+	// Add the samples to the stream, angled in L-R space
+	const int sample_end = requested_frames * 2 - 1;
+	for (int i = 0; i < sample_end; i += 2) {
+		float sample = GetSample(ram);
+		sample *= GetVolScalar(vol_scalars);
+		stream[i] += sample * pan_scalars[pan_position].left;
+		stream[i + 1] += sample * pan_scalars[pan_position].right;
 	}
-	void WriteRampRate(Bit8u val) {
-		RampRate = val;
-		double frameadd = (double)(RampRate & 63)/(double)(1 << (3*(val >> 6)));
-		double realadd = (frameadd*(double)myGUS.basefreq/(double)GUS_RATE) * (double)(1 << RAMP_FRACT);
-		RampAdd = (Bit32u)realadd;
+	// Keep track of how many ms this voice has generated
+	Is8Bit() ? generated_8bit_ms++ : generated_16bit_ms++;
+}
+
+int32_t Voice::PopWavePos()
+{
+	const int32_t pos = wave_ctrl.pos;
+	IncrementCtrlPos(wave_ctrl, CheckWaveRolloverCondition());
+	return pos;
```

Comment:
```suggestion
	IncrementCtrlPos(wave_ctrl, CheckWaveRolloverCondition());
	return wave_ctrl.pos;
```
? The value is already an `int32_t`.

Reply:
I can document it better; the distinction is we need to return the "current" position, not the incremented position (which is used next-go-round).